rust/otap-dataflow/crates/otap/src/otel_receiver.rs

```diff
@@ -0,0 +1,994 @@
+                            AdmitDecision::Reject { message } => {
+                                // Breaker/policy: reject this stream immediately.
+                                respond_with_error(
+                                    respond,
+                                    Status::unavailable(message),
+                                    &router.request_accept_header,
+                                );
+                            }
+                        }
+                    }
+                    Some(Err(err)) => return Err(err),
+                    None => accepting = false,
+                }
+            }
+
+            // Join completed stream tasks
+            maybe_done = stream_tasks.join_next(), if !stream_tasks.is_empty() => {
+                if let Some(Err(join_err)) = maybe_done {
+                    if log::log_enabled!(log::Level::Debug) {
+                        log::debug!("stream task join error: {join_err}");
+                    }
+                }
+            }
+
+            keepalive_result = async {
+                if let Some(ka) = keepalive.as_mut() {
+                    ka.poll_tick().await
+                } else {
+                    unreachable!("keepalive polled without being armed");
+                }
+            }, if keepalive_armed => {
+                match keepalive_result {
+                    Ok(()) => {}
+                    Err(err) => {
+                        if log::log_enabled!(log::Level::Debug) {
+                            log::debug!("H2 keepalive failed: {err}");
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+
+        // Exit when no more streams will arrive and all tasks are done
+        if !accepting && stream_tasks.is_empty() {
+            break;
+        }
+    }
+
+    // Drain in-flight stream tasks
+    while let Some(res) = stream_tasks.join_next().await {
+        if let Err(join_err) = res {
+            if log::log_enabled!(log::Level::Debug) {
+                log::debug!("stream task join error: {join_err}");
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Tracks whether a connection needs a HTTP/2 PING/PONG to keep the client alive.
+struct Http2Keepalive {
+    ping_pong: PingPong,
+    interval: Duration,
+    timeout: Duration,
+    sleep: Option<Pin<Box<Sleep>>>,
+}
+
+impl Http2Keepalive {
+    fn new(
+        ping_pong: Option<PingPong>,
+        interval: Option<Duration>,
+        timeout: Option<Duration>,
+    ) -> Option<Self> {
+        let (ping_pong, interval, timeout) = match (ping_pong, interval, timeout) {
+            (Some(ping_pong), Some(interval), Some(timeout)) if !interval.is_zero() => {
+                (ping_pong, interval, timeout)
+            }
+            _ => return None,
+        };
+        Some(Self {
+            ping_pong,
+            interval,
+            timeout,
+            sleep: None,
+        })
```

Comment:
nit: could this be simplified as follows?
```suggestion
        let interval = interval?;
        if interval.is_zero() {
            return None
        }

        Some(Self {
            ping_pong: ping_pong?,
            interval,
            timeout: timeout?,
            sleep: None,
        })
```